File: cache/java/src/main/java/com/example/cache/Cache.java
package com.example.cache;

import com.example.cache.eviction.EvictionPolicy;
import com.example.cache.exception.InvalidCapacityException;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Generic in-memory cache with pluggable eviction.
 *
 * <p>The cache stores values in a HashMap and delegates all admission/access/eviction ordering to an
 * EvictionPolicy. get/put/size/containsKey are guarded by one ReentrantLock. This coarse lock keeps
 * the MVP easy to reason about: the value map and policy metadata always change atomically together,
 * so size never exceeds capacity even when many threads hammer put/get concurrently.
 */
public class Cache<K, V> {

    private final int capacity;
    private final EvictionPolicy<K> evictionPolicy;
    private final Map<K, V> values = new HashMap<>();
    private final ReentrantLock lock = new ReentrantLock();

    public Cache(int capacity, EvictionPolicy<K> evictionPolicy) {
        if (capacity <= 0) {
            throw new InvalidCapacityException("Capacity must be positive");
        }
        this.capacity = capacity;
        this.evictionPolicy = evictionPolicy;
    }

    /** Return Optional.empty() on miss; a hit updates recency/frequency through the policy. */
    public Optional<V> get(K key) {
        lock.lock();
        try {
            if (!values.containsKey(key)) {
                return Optional.empty();
            }
            evictionPolicy.keyAccessed(key);
            return Optional.ofNullable(values.get(key));
        } finally {
            lock.unlock();
        }
    }

    /** Insert or update. If full, evict one victim selected by the policy before admitting a new key. */
    public void put(K key, V value) {
        lock.lock();
        try {
            if (values.containsKey(key)) {
                values.put(key, value);
                evictionPolicy.keyAccessed(key); // update refreshes LRU and increments LFU frequency
                return;
            }
            if (values.size() >= capacity) {
                K victim = evictionPolicy.evictKey();
                if (victim != null) {
                    values.remove(victim);
                }
            }
            values.put(key, value);
            evictionPolicy.keyInserted(key);
        } finally {
            lock.unlock();
        }
    }

    public int size() {
        lock.lock();
        try {
            return values.size();
        } finally {
            lock.unlock();
        }
    }

    public int capacity() {
        return capacity;
    }

    /** Membership check intentionally does not update recency/frequency; only get/put are accesses. */
    public boolean containsKey(K key) {
        lock.lock();
        try {
            return values.containsKey(key);
        } finally {
            lock.unlock();
        }
    }
}

File: cache/python/cache/cache.py
"""Thread-safe generic in-memory cache with pluggable eviction."""

from __future__ import annotations

import threading
from typing import Generic, Hashable, TypeVar

from .exceptions import InvalidCapacityError
from .policies import EvictionPolicy

K = TypeVar("K", bound=Hashable)
V = TypeVar("V")
_MISSING = object()


class Cache(Generic[K, V]):
    """Cache service using Strategy for eviction and an RLock for concurrency.

    A single lock protects both the value dict and the policy metadata. That coarse-grained model is
    deliberate for a machine-coding MVP: every get/put is atomic, policy state cannot drift away from
    stored values, and size never exceeds capacity under concurrent access.
    """

    def __init__(self, capacity: int, eviction_policy: EvictionPolicy[K]) -> None:
        if capacity <= 0:
            raise InvalidCapacityError("Capacity must be positive")
        self._capacity = capacity
        self._policy = eviction_policy
        self._values: dict[K, V] = {}
        self._lock = threading.RLock()

    def get(self, key: K) -> V | None:
        """Return the value or None on miss; hits update recency/frequency."""
        with self._lock:
            value = self._values.get(key, _MISSING)
            if value is _MISSING:
                return None
            self._policy.key_accessed(key)
            return value  # type: ignore[return-value]

    def put(self, key: K, value: V) -> None:
        """Insert/update a value, evicting through the policy when full."""
        with self._lock:
            if key in self._values:
                self._values[key] = value
                self._policy.key_accessed(key)
                return
            if len(self._values) >= self._capacity:
                victim = self._policy.evict_key()
                if victim is not None:
                    self._values.pop(victim, None)
            self._values[key] = value
            self._policy.key_inserted(key)

    def size(self) -> int:
        with self._lock:
            return len(self._values)

    @property
    def capacity(self) -> int:
        return self._capacity

    def contains_key(self, key: K) -> bool:
        """Membership check does not count as an access; only get/put alter eviction state."""
        with self._lock:
            return key in self._values
